### PR TITLE
gh-102837: few coverage nitpicks for the math module

### PR DESCRIPTION
**Pull request by skirpichev**
- input checks for math_1(L989), math_1a(L1023), math_2(L1064,L1071), hypot(L2682), log(L2307), ldexp(L2168), ceil(L1165), floor(L1236,L1239) and dist(L2587,L2588,L2628).
- drop inaccessible "if" branch (L3518) in perm_comb_small()
- improve fsum coverage for exceptional cases (L1433,L1438,L1451,L1497), ditto fmod(L2378)
- rewrite modf to fix inaccessible case(L2229), ditto for pow(L2988)
    
(all line numbers are wrt the main branch at 5e6661bce9)

* Issue: gh-102837

### CODE REVIEW

**Comment by skirpichev** [comment]
Note, c.f. #102067 for the cmath module - there are some uncovered lines/branches.

**Comment by skirpichev** [comment]
@AlexWaygood, do you think this requires an issue?

**Comment by AlexWaygood** [comment]
> @AlexWaygood, do you think this requires an issue?

Our policy is generally to require issues for all changes, unless they're _extremely_ trivial (such as typo fixes).

**Comment by skirpichev** [comment]
On Wed, Mar 08, 2023 at 06:21:43AM -0800, Alex Waygood wrote:
>    Our policy is generally to require issues for all changes, unless they're
>    extremely trivial (such as typo fixes).

Ok.  I thought this is trivial:)

File: Modules/mathmodule.c [hunk 1]
```diff
@@ -3479,10 +3464,6 @@ static const uint8_t factorial_trailing_zeros[] = {
 static PyObject *
 perm_comb_small(unsigned long long n, unsigned long long k, int iscomb)
 {
-    if (k == 0) {
```
Comment (hauntsaninja) [inline Modules/mathmodule.c:3483]: I don't think we should remove this just because it's unreachable from test cases. It's not obvious to me that `perm_comb_small` is correct without it, given potential for infinite recursion or infinite unsigned looping. If you feel strongly about removing it, probably should at least be an assert.

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -2949,7 +2934,7 @@ math_pow_impl(PyObject *module, double x, double y)
             else /* y < 0. */
                 r = odd_y ? copysign(0., x) : 0.;
         }
-        else if (Py_IS_INFINITY(y)) {
+        else { /* Py_IS_INFINITY(y) */
```
Comment (hauntsaninja) [inline Modules/mathmodule.c]: Make this an assert?

File: Modules/mathmodule.c [hunk 3]
```diff
@@ -961,24 +961,16 @@ math_1(PyObject *arg, double (*func) (double), int can_overflow)
...
+            errno = ERANGE;
         else
-            PyErr_SetString(PyExc_ValueError,
-                            "math domain error"); /* singularity */
-        return NULL;
+            errno = EDOM;
     }
-    if (Py_IS_FINITE(r) && errno && is_error(r))
-        /* this branch unnecessary on most platforms */
+    if (errno && is_error(r))
```
Comment (hauntsaninja) [inline Modules/mathmodule.c]: I don't understand this well enough to review it — do you know what situation / platform the left hand side code is worried about?

**Review by hauntsaninja — COMMENTED** [review]
Thanks!

File: Modules/mathmodule.c (hunk 3 above)
Comment (skirpichev) [reply Modules/mathmodule.c]: Well, this is for some broken libm.  de4296281a shows no glue on which one (if it does exists).  I'll revert this part, however.

File: Modules/mathmodule.c (hunk 1 above)
Comment (skirpichev) [reply Modules/mathmodule.c:3483]: > unreachable from test cases

Not just from tests...

> probably should at least be an assert

Ok.

File: Modules/mathmodule.c (hunk 3 above)
Comment (skirpichev) [reply Modules/mathmodule.c]: Oh, I got it.  After 9c32ae0, ``r`` could be infinite (or nan) at this point.  But comment should be restored, of course.

BTW, in some sense 9c32ae0 could be viewed as an unrelated code refactoring (should be helpful for #101410, for example), yet that improved coverage on one line.  Do you suggest to revert this commit?

Comment (hauntsaninja) [reply Modules/mathmodule.c]: Yeah, if you revert I can merge, otherwise will need to wait for someone else to confirm correctness

File: Modules/mathmodule.c [hunk 4]
```diff
@@ -1207,10 +1194,11 @@ math_floor(PyObject *module, PyObject *number)
         }
         if (PyErr_Occurred())
             return NULL;
-        x = PyFloat_AsDouble(number);
-        if (x == -1.0 && PyErr_Occurred())
-            return NULL;
     }
+    double x = PyFloat_AsDouble(number);
```
Comment (hauntsaninja) [inline Modules/mathmodule.c]: I missed this on my first review, but PyFloat_AsDouble is not the same as PyFloat_AS_DOUBLE, since the latter is faster. Maybe we should make math_ceil match the math_floor logic (best done in separate PR)?

Comment (hauntsaninja) [reply Modules/mathmodule.c]: Yup, this was an optimisation added in https://github.com/python/cpython/pull/21072. I'll open a PR making math_ceil match the logic in math_floor

Comment (hauntsaninja) [reply Modules/mathmodule.c]: Opened https://github.com/python/cpython/pull/108801

File: Modules/mathmodule.c (hunk 3 above)
Comment (skirpichev) [reply Modules/mathmodule.c]: @mdickinson, maybe you could review that part (second commit, https://github.com/python/cpython/commit/9c32ae022f07ea90892175a2d7f5c777621be0e0)?

File: Modules/mathmodule.c (hunk 4 above)
Comment (skirpichev) [reply Modules/mathmodule.c]: Thanks, I've reverted the math_floor() change and added new tests to cover L1239.  (And similar tests for ceil() to help you bring coverage back in #108801.)

**Comment by skirpichev** [comment]
Last commit will cover rest of math_floor()/ceil().

File: Modules/mathmodule.c (hunk 3 above)
Comment (skirpichev) [reply Modules/mathmodule.c]: Ok, @hauntsaninja, https://github.com/python/cpython/commit/9c32ae022f07ea90892175a2d7f5c777621be0e0 was reverted.  Perhaps, this can go to other PR...

The PR description was updated (maybe you can use this for a commit message).

**Review by hauntsaninja — APPROVED** [review]
Thanks for the follow ups! :-)

**Comment by skirpichev** [comment]
Next pr: #110000

File: Modules/mathmodule.c [hunk 5]
```diff
@@ -2195,12 +2195,10 @@ math_modf_impl(PyObject *module, double x)
...
-    if (!Py_IS_FINITE(x)) {
-        if (Py_IS_INFINITY(x))
-            return Py_BuildValue("(dd)", copysign(0., x), x);
-        else if (Py_IS_NAN(x))
-            return Py_BuildValue("(dd)", x, x);
-    }
+    if (Py_IS_INFINITY(x))
+        return Py_BuildValue("(dd)", copysign(0., x), x);
+    else if (Py_IS_NAN(x))
+        return Py_BuildValue("(dd)", x, x);
```
Comment (serhiy-storchaka) [inline Modules/mathmodule.c:2201]: The old code was more efficient in common case: only one check `Py_IS_FINITE()` instead of two checks `Py_IS_INFINITY()` and `Py_IS_NAN()`. It may be not so important in this case, fut I would prefer to revert this change, just to be consistent with the rest of the code.

File: Modules/mathmodule.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/mathmodule.c:3483]: I guarantee that it is never 0 in the current code. In the outermost call k > 1, and in recursive calls k can be 0 only if k is 0 or 1 in the upper call, but the case k == 1 is handled before recursion by setting `fast_comb_limits2[1]` and `fast_perm_limits[1]` to `ULLONG_MAX`.

So the runtime check can be removed, but the assert can help if the code be changed in future.

File: Modules/mathmodule.c (hunk 5 above)
Comment (skirpichev) [reply Modules/mathmodule.c:2201]: Indeed, this is less efficient for ``Py_IS_FINITE()`` case.  But I think we could rewrite this with ``else`` clause instead of ``else if (Py_IS_NAN(x))``.

Comment (serhiy-storchaka) [reply Modules/mathmodule.c:2201]: Maybe. If neither of them has bugs.

Comment (skirpichev) [reply Modules/mathmodule.c:2201]: I think we could be safe that no libm is so broken in 2023.  Also it's already assumed in many places across the module (e.g. in ``m_tgamma()``, ``m_lgamma()``, ``m_log2()``, etc).

Comment (skirpichev) [reply Modules/mathmodule.c:2201]: Thanks.  Patch added to #110000.
